<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>BenchmarkTask.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-app</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.app.cli.variant.annotation</a> &gt; <span class="el_source">BenchmarkTask.java</span></div><h1>BenchmarkTask.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.app.cli.variant.annotation;

import org.apache.commons.lang3.tuple.Pair;
import org.opencb.biodata.models.variant.Variant;
import org.opencb.biodata.models.variant.avro.ConsequenceType;
import org.opencb.biodata.models.variant.avro.SequenceOntologyTerm;
import org.opencb.biodata.models.variant.avro.VariantAnnotation;
import org.opencb.biodata.tools.sequence.FastaIndexManager;
import org.opencb.cellbase.core.variant.annotation.VariantAnnotator;
import org.opencb.commons.run.ParallelTaskRunner;
import org.rocksdb.RocksDBException;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Created by fjlopez on 07/04/16.
 */
public class BenchmarkTask implements
        ParallelTaskRunner.TaskWithException&lt;VariantAnnotation, Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt;, Exception&gt; {

    private FastaIndexManager fastaIndexManager;
    private static final String VARIANT_STRING_PATTERN = &quot;[ACGT]*&quot;;
//    private static final String VARIANT_STRING_PATTERN = &quot;([ACGT]*)|(&lt;CNV&gt;)|(&lt;INV&gt;)|(&lt;DEL&gt;)|(&lt;INS&gt;)|(&lt;DUP:TANDEM&gt;)&quot;;
    private VariantAnnotator variantAnnotator;

<span class="nc" id="L29">    public BenchmarkTask(VariantAnnotator variantAnnotator, FastaIndexManager fastaIndexManager) {</span>
<span class="nc" id="L30">        this.variantAnnotator = variantAnnotator;</span>
<span class="nc" id="L31">        this.fastaIndexManager = fastaIndexManager;</span>
<span class="nc" id="L32">    }</span>

    public void pre() {
<span class="nc" id="L35">        variantAnnotator.open();</span>
<span class="nc" id="L36">    }</span>

    public List&lt;Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt;&gt; apply(List&lt;VariantAnnotation&gt; batch)
            throws Exception {
        // VEP format does not include the reference allele. It's needed for the benchmark if the cache is activated -
        // otherwise VariantAnnotationCalculator wont find most of the variants in the variation collection
<span class="nc" id="L42">        fixReference(batch);</span>
<span class="nc" id="L43">        removeInvalidVariants(batch);</span>
<span class="nc" id="L44">        List&lt;Variant&gt; cellBaseBatch = createEmptyVariantList(batch);</span>
<span class="nc" id="L45">        variantAnnotator.run(cellBaseBatch);</span>
<span class="nc" id="L46">        List&lt;Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt;&gt; comparisonResultList = new ArrayList&lt;&gt;();</span>
<span class="nc bnc" id="L47" title="All 2 branches missed.">        for (int i = 0; i &lt; batch.size(); i++) {</span>
            // Variants such as MT:453:TTT:ATT are skipped for the benchmark - will not have CellBase annotation
            // compatible with VEP annotation and therefore consequenceTypeList = null
<span class="nc bnc" id="L50" title="All 4 branches missed.">            if (batch.get(i).getConsequenceTypes() != null &amp;&amp; batch.get(i).getConsequenceTypes().size() &gt; 0</span>
<span class="nc bnc" id="L51" title="All 2 branches missed.">                    &amp;&amp; cellBaseBatch.get(i).getAnnotation().getConsequenceTypes() != null</span>
<span class="nc bnc" id="L52" title="All 2 branches missed.">                    &amp;&amp; cellBaseBatch.get(i).getAnnotation().getConsequenceTypes().size() &gt; 0) {</span>
<span class="nc" id="L53">                Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt; comparisonResult = compare(batch.get(i),</span>
<span class="nc" id="L54">                        cellBaseBatch.get(i).getAnnotation());</span>
<span class="nc" id="L55">                comparisonResult.getLeft().setVariantAnnotation(batch.get(i));</span>
<span class="nc" id="L56">                comparisonResult.getRight().setVariantAnnotation(cellBaseBatch.get(i).getAnnotation());</span>
<span class="nc" id="L57">                comparisonResultList.add(comparisonResult);</span>
            }
        }
<span class="nc" id="L60">        return comparisonResultList;</span>
    }

    private void fixReference(List&lt;VariantAnnotation&gt; variantAnnotationList) throws RocksDBException {
<span class="nc bnc" id="L64" title="All 2 branches missed.">        for (VariantAnnotation variantAnnotation : variantAnnotationList) {</span>
<span class="nc bnc" id="L65" title="All 4 branches missed.">            if (!variantAnnotation.getReference().isEmpty() &amp;&amp; !variantAnnotation.getReference().equals(&quot;-&quot;)) {</span>
<span class="nc" id="L66">                variantAnnotation.setReference(fastaIndexManager.query(variantAnnotation.getChromosome(),</span>
<span class="nc" id="L67">                        variantAnnotation.getStart(), variantAnnotation.getStart()</span>
<span class="nc" id="L68">                                + variantAnnotation.getReference().length() - 1));</span>
            }
<span class="nc" id="L70">        }</span>
<span class="nc" id="L71">    }</span>

    private void removeInvalidVariants(List&lt;VariantAnnotation&gt; variantAnnotationList) {
<span class="nc" id="L74">        int i = 0;</span>
<span class="nc bnc" id="L75" title="All 2 branches missed.">        while (i &lt; variantAnnotationList.size()) {</span>
<span class="nc bnc" id="L76" title="All 2 branches missed.">            if (isValid(variantAnnotationList.get(i))) {</span>
<span class="nc" id="L77">                i++;</span>
            } else {
<span class="nc" id="L79">                variantAnnotationList.remove(i);</span>
            }
        }
<span class="nc" id="L82">    }</span>

    /**
     * Checks whether a variant is valid.
     *
     * @param variantAnnotation Variant object to be checked.
     * @return   true/false depending on whether 'variant' does contain valid values. Currently just a simple check of
     * reference/alternate attributes being strings of [A,C,G,T] of length &gt;= 0 is performed to detect cases such as
     * 19:13318673:(CAG)4:(CAG)5 which are not currently supported by CellBase. Ref and alt alleles must be different
     * as well for the variant to be valid. Functionality of the method may be improved in the future.
     */
    private boolean isValid(VariantAnnotation variantAnnotation) {
<span class="nc bnc" id="L94" title="All 2 branches missed.">        return (variantAnnotation.getAlternate().matches(VARIANT_STRING_PATTERN)</span>
//                &amp;&amp; variantAnnotation.getReference().matches(VARIANT_STRING_PATTERN)
<span class="nc bnc" id="L96" title="All 2 branches missed.">                &amp;&amp; !variantAnnotation.getAlternate().equals(variantAnnotation.getReference()));</span>
    }

    private Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt; compare(VariantAnnotation variant1, VariantAnnotation variant2) {
<span class="nc" id="L100">        Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt; result</span>
<span class="nc" id="L101">                = Pair.of(new VariantAnnotationDiff(), new VariantAnnotationDiff());</span>
<span class="nc" id="L102">        compareSequenceOntologyTerms(result, variant1.getConsequenceTypes(),</span>
<span class="nc" id="L103">                variant2.getConsequenceTypes());</span>

<span class="nc" id="L105">        return result;</span>
    }

    private void compareSequenceOntologyTerms(Pair&lt;VariantAnnotationDiff, VariantAnnotationDiff&gt; result,
                                              List&lt;ConsequenceType&gt; consequenceTypeList1,
                                              List&lt;ConsequenceType&gt; consequenceTypeList2) {
<span class="nc" id="L111">        Set&lt;SequenceOntologyTermComparisonObject&gt; sequenceOntologySet1 = getSequenceOntologySet(consequenceTypeList1);</span>
<span class="nc" id="L112">        Set&lt;SequenceOntologyTermComparisonObject&gt; sequenceOntologySet2 = getSequenceOntologySet(consequenceTypeList2);</span>
<span class="nc" id="L113">        Set&lt;SequenceOntologyTermComparisonObject&gt; sequenceOntologySet1bak = new HashSet&lt;&gt;(sequenceOntologySet1);</span>
<span class="nc" id="L114">        sequenceOntologySet1.removeAll(sequenceOntologySet2);</span>
<span class="nc" id="L115">        sequenceOntologySet2.removeAll(sequenceOntologySet1bak);</span>
<span class="nc bnc" id="L116" title="All 2 branches missed.">        if (sequenceOntologySet1.size() &gt; 0) {</span>
<span class="nc" id="L117">            result.getLeft().setSequenceOntology(new ArrayList(sequenceOntologySet1));</span>
        }
<span class="nc bnc" id="L119" title="All 2 branches missed.">        if (sequenceOntologySet2.size() &gt; 0) {</span>
<span class="nc" id="L120">            result.getRight().setSequenceOntology(new ArrayList(sequenceOntologySet2));</span>
        }
<span class="nc" id="L122">    }</span>

    private Set&lt;SequenceOntologyTermComparisonObject&gt; getSequenceOntologySet(List&lt;ConsequenceType&gt; consequenceTypeList) {
<span class="nc bnc" id="L125" title="All 2 branches missed.">        if (consequenceTypeList != null) {</span>
<span class="nc" id="L126">            Set&lt;SequenceOntologyTermComparisonObject&gt; set = new HashSet&lt;&gt;(consequenceTypeList.size());</span>
<span class="nc bnc" id="L127" title="All 2 branches missed.">            for (ConsequenceType consequenceType : consequenceTypeList) {</span>
<span class="nc bnc" id="L128" title="All 2 branches missed.">                for (SequenceOntologyTerm sequenceOntologyTerm : consequenceType.getSequenceOntologyTerms()) {</span>
//                    // Expected many differences depending on the regulatory source databases used by the annotators.
//                    // Better skip regulatory_region_variant annotations
//                    if (!(sequenceOntologyTerm.getName().equals(VariantAnnotationUtils.REGULATORY_REGION_VARIANT)
//                            || sequenceOntologyTerm.getName().equals(VariantAnnotationUtils.TF_BINDING_SITE_VARIANT))) {
<span class="nc" id="L133">                    set.add(new SequenceOntologyTermComparisonObject(consequenceType.getEnsemblTranscriptId(),</span>
                            sequenceOntologyTerm));
//                    }
<span class="nc" id="L136">                }</span>
<span class="nc" id="L137">            }</span>

<span class="nc" id="L139">            return set;</span>
        } else {
<span class="nc" id="L141">            return null;</span>
        }
    }

    private List&lt;Variant&gt; createEmptyVariantList(List&lt;VariantAnnotation&gt; variantAnnotationList) {
<span class="nc" id="L146">        List&lt;Variant&gt; newVariantList = new ArrayList&lt;&gt;(variantAnnotationList.size());</span>
<span class="nc bnc" id="L147" title="All 2 branches missed.">        for (VariantAnnotation variantAnnotation : variantAnnotationList) {</span>
<span class="nc" id="L148">            Variant variant = new Variant(variantAnnotation.getChromosome(), variantAnnotation.getStart(),</span>
<span class="nc" id="L149">                    variantAnnotation.getEnd(), variantAnnotation.getReference(), variantAnnotation.getAlternate());</span>
<span class="nc" id="L150">            variant.resetType();</span>
<span class="nc" id="L151">            variant.resetLength();</span>
<span class="nc" id="L152">            newVariantList.add(variant);</span>
<span class="nc" id="L153">        }</span>

<span class="nc" id="L155">        return newVariantList;</span>
    }

    public void post() {
<span class="nc" id="L159">        variantAnnotator.close();</span>
<span class="nc" id="L160">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>