<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>CellBaseNormalizerSequenceAdaptor.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-core</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.core.variant.annotation</a> &gt; <span class="el_source">CellBaseNormalizerSequenceAdaptor.java</span></div><h1>CellBaseNormalizerSequenceAdaptor.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.core.variant.annotation;

import org.apache.commons.lang3.StringUtils;
import org.opencb.biodata.models.core.GenomeSequenceFeature;
import org.opencb.biodata.models.core.Region;
import org.opencb.biodata.tools.sequence.SequenceAdaptor;
import org.opencb.cellbase.core.api.GenomeDBAdaptor;
import org.opencb.commons.datastore.core.QueryOptions;
import org.opencb.commons.datastore.core.QueryResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CellBaseNormalizerSequenceAdaptor implements SequenceAdaptor {
    private final GenomeDBAdaptor genomeDBAdaptor;
<span class="nc" id="L15">    private final Logger logger = LoggerFactory.getLogger(this.getClass());</span>

<span class="nc" id="L17">    public CellBaseNormalizerSequenceAdaptor(GenomeDBAdaptor genomeDBAdaptor) {</span>
<span class="nc" id="L18">        this.genomeDBAdaptor = genomeDBAdaptor;</span>
<span class="nc" id="L19">    }</span>

    /**
     * Returns sequence from contig &quot;contig&quot; in the range [start, end] (1-based, both inclusive).
     * For corner cases mimics the behaviour of the org.opencb.biodata.tools.sequence.SamtoolsFastaIndex with one
     * difference. If chromosome does not exist, start is under the left bound, start AND end are out of the right
     * bound, then a RunTime exception will be thrown. HOWEVER: if start is within the bounds BUT end is out of the
     * right bound, then THIS implementaiton will return available nucleotides while SamtoolsFastaIndex will keep
     * returning the exception.
     * @param contig
     * @param start
     * @param end
     * @return String containing the sequence of contig &quot;contig&quot; in the range [start, end] (1-based, both inclusive).
     * Throws RunTimeException if contig does not exist, or start is under the left bound, or start AND end are out of the right
     * bound. If start is within the bounds BUT end is out of the
     * right bound, then THIS implementaiton will return available nucleotides while SamtoolsFastaIndex will keep
     * returning the exception.
     * @throws Exception
     * @throws RuntimeException
     */
    @Override
    public String query(String contig, int start, int end) throws Exception {
<span class="nc" id="L41">        Region region = new Region(contig, start, end);</span>
<span class="nc" id="L42">        QueryResult&lt;GenomeSequenceFeature&gt; queryResult</span>
<span class="nc" id="L43">                 = genomeDBAdaptor.getSequence(region, QueryOptions.empty());</span>

        // This behaviour mimics the behaviour of the org.opencb.biodata.tools.sequence.SamtoolsFastaIndex with one
        // difference. If contig does not exist, start is under the left bound, start AND end are out of the right
        // bound, then a RunTime exception will be thrown. HOWEVER: if start is within the bounds BUT end is out of the
        // right bound, then THIS implementaiton will return available nucleotides while SamtoolsFastaIndex will keep
        // returning the exception.
<span class="nc bnc" id="L50" title="All 4 branches missed.">        if (queryResult.getResult().size() &gt; 0 &amp;&amp; StringUtils.isNotBlank(queryResult.getResult().get(0).getSequence())) {</span>
<span class="nc bnc" id="L51" title="All 2 branches missed.">            if (queryResult.getResult().get(0).getSequence().length() &lt; (end - start + 1)) {</span>
<span class="nc" id="L52">                logger.warn(&quot;End coordinate out of the right bound. Returning available nucleotides.&quot;);</span>
            }
<span class="nc" id="L54">            return queryResult.getResult().get(0).getSequence();</span>
        } else {
<span class="nc" id="L56">            throw new RuntimeException(&quot;Unable to find entry for &quot; + region.toString());</span>
        }
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>