<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ConsequenceTypeCNVGainCalculator.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-core</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.core.variant.annotation</a> &gt; <span class="el_source">ConsequenceTypeCNVGainCalculator.java</span></div><h1>ConsequenceTypeCNVGainCalculator.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.core.variant.annotation;

import org.opencb.biodata.models.core.Gene;
import org.opencb.biodata.models.core.Transcript;
import org.opencb.biodata.models.variant.Variant;
import org.opencb.biodata.models.variant.avro.ConsequenceType;
import org.opencb.commons.datastore.core.QueryOptions;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by fjlopez on 31/05/16.
 */
public class ConsequenceTypeCNVGainCalculator extends ConsequenceTypeGenericRegionCalculator {

<span class="nc" id="L17">    public ConsequenceTypeCNVGainCalculator() { }</span>

    public List&lt;ConsequenceType&gt; run(Variant inputVariant, List&lt;Gene&gt; geneList, boolean[] overlapsRegulatoryRegion,
                                     QueryOptions queryOptions) {
<span class="nc" id="L21">        parseQueryParam(queryOptions);</span>
<span class="nc" id="L22">        List&lt;ConsequenceType&gt; consequenceTypeList = new ArrayList&lt;&gt;();</span>
<span class="nc" id="L23">        variant = inputVariant;</span>
<span class="nc" id="L24">        variantEnd = getEnd(cnvExtraPadding);</span>
<span class="nc" id="L25">        variantStart = getStart(cnvExtraPadding);</span>
//        isBigDeletion = ((variantEnd - variantStart) &gt; BIG_VARIANT_SIZE_THRESHOLD);
<span class="nc" id="L27">        boolean isIntergenic = true;</span>
<span class="nc bnc" id="L28" title="All 2 branches missed.">        for (Gene currentGene : geneList) {</span>
<span class="nc" id="L29">            gene = currentGene;</span>
<span class="nc bnc" id="L30" title="All 2 branches missed.">            for (Transcript currentTranscript : gene.getTranscripts()) {</span>
<span class="nc bnc" id="L31" title="All 6 branches missed.">                isIntergenic = isIntergenic &amp;&amp; (variantEnd &lt; currentTranscript.getStart() || variantStart &gt; currentTranscript.getEnd());</span>
<span class="nc" id="L32">                transcript = currentTranscript;</span>
<span class="nc" id="L33">                consequenceType = new ConsequenceType();</span>
<span class="nc" id="L34">                consequenceType.setGeneName(gene.getName());</span>
<span class="nc" id="L35">                consequenceType.setEnsemblGeneId(gene.getId());</span>
<span class="nc" id="L36">                consequenceType.setEnsemblTranscriptId(transcript.getId());</span>
<span class="nc" id="L37">                consequenceType.setStrand(transcript.getStrand());</span>
<span class="nc" id="L38">                consequenceType.setBiotype(transcript.getBiotype());</span>
<span class="nc bnc" id="L39" title="All 2 branches missed.">                consequenceType.setTranscriptAnnotationFlags(transcript.getAnnotationFlags() != null</span>
<span class="nc" id="L40">                        ? new ArrayList&lt;&gt;(transcript.getAnnotationFlags()) : null);</span>
<span class="nc" id="L41">                SoNames.clear();</span>

<span class="nc bnc" id="L43" title="All 2 branches missed.">                if (transcript.getStrand().equals(&quot;+&quot;)) {</span>
                    // whole transcript affected
<span class="nc bnc" id="L45" title="All 4 branches missed.">                    if (variantStart &lt;= transcript.getStart() &amp;&amp; variantEnd &gt;= transcript.getEnd()) {</span>
<span class="nc" id="L46">                        SoNames.add(VariantAnnotationUtils.TRANSCRIPT_AMPLIFICATION);</span>
<span class="nc" id="L47">                        consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L48">                        consequenceTypeList.add(consequenceType);</span>
<span class="nc bnc" id="L49" title="All 2 branches missed.">                    } else if (regionsOverlap(transcript.getStart(), transcript.getEnd(), variantStart, variantEnd)) {</span>
<span class="nc" id="L50">                        solvePositiveTranscript(consequenceTypeList);</span>
                    } else {
<span class="nc" id="L52">                        solveTranscriptFlankingRegions(VariantAnnotationUtils.UPSTREAM_GENE_VARIANT,</span>
                                VariantAnnotationUtils.DOWNSTREAM_GENE_VARIANT);
<span class="nc bnc" id="L54" title="All 2 branches missed.">                        if (SoNames.size() &gt; 0) { // Variant does not overlap gene region, just may have upstream/downstream annotations</span>
<span class="nc" id="L55">                            consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L56">                            consequenceTypeList.add(consequenceType);</span>
                        }
                    }
                } else {
<span class="nc bnc" id="L60" title="All 4 branches missed.">                    if (variantStart &lt;= transcript.getStart() &amp;&amp; variantEnd &gt;= transcript.getEnd()) { // whole trans. affected</span>
<span class="nc" id="L61">                        SoNames.add(VariantAnnotationUtils.TRANSCRIPT_AMPLIFICATION);</span>
<span class="nc" id="L62">                        consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L63">                        consequenceTypeList.add(consequenceType);</span>
<span class="nc bnc" id="L64" title="All 2 branches missed.">                    } else if (regionsOverlap(transcript.getStart(), transcript.getEnd(), variantStart, variantEnd)) {</span>
<span class="nc" id="L65">                        solveNegativeTranscript(consequenceTypeList);</span>
                    } else {
<span class="nc" id="L67">                        solveTranscriptFlankingRegions(VariantAnnotationUtils.DOWNSTREAM_GENE_VARIANT,</span>
                                VariantAnnotationUtils.UPSTREAM_GENE_VARIANT);
<span class="nc bnc" id="L69" title="All 2 branches missed.">                        if (SoNames.size() &gt; 0) { // Variant does not overlap gene region, just has upstream/downstream annotations</span>
<span class="nc" id="L70">                            consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L71">                            consequenceTypeList.add(consequenceType);</span>
                        }
                    }
                }
<span class="nc" id="L75">            }</span>
<span class="nc" id="L76">        }</span>

<span class="nc" id="L78">        solveIntergenic(consequenceTypeList, isIntergenic);</span>
<span class="nc" id="L79">        solveRegulatoryRegions(overlapsRegulatoryRegion, consequenceTypeList);</span>
<span class="nc" id="L80">        return consequenceTypeList;</span>
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>