<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>ConsequenceTypeMNVCalculator.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-core</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.core.variant.annotation</a> &gt; <span class="el_source">ConsequenceTypeMNVCalculator.java</span></div><h1>ConsequenceTypeMNVCalculator.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.core.variant.annotation;

import org.opencb.biodata.models.core.Gene;
import org.opencb.biodata.models.core.Transcript;
import org.opencb.biodata.models.variant.Variant;
import org.opencb.biodata.models.variant.avro.ConsequenceType;
import org.opencb.cellbase.core.api.GenomeDBAdaptor;
import org.opencb.commons.datastore.core.QueryOptions;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by fjlopez on 15/03/17.
 */
public class ConsequenceTypeMNVCalculator extends ConsequenceTypeGenericRegionCalculator {
//    private int variantStart;
//    private int variantEnd;
//    private GenomeDBAdaptor genomeDBAdaptor;

//    private static final int BIG_VARIANT_SIZE_THRESHOLD = 50;


<span class="nc" id="L24">    public ConsequenceTypeMNVCalculator(GenomeDBAdaptor genomeDBAdaptor) {</span>
<span class="nc" id="L25">        this.genomeDBAdaptor = genomeDBAdaptor;</span>
<span class="nc" id="L26">    }</span>

    public List&lt;ConsequenceType&gt; run(Variant inputVariant, List&lt;Gene&gt; geneList, boolean[] overlapsRegulatoryRegion,
                                     QueryOptions queryOptions) {
<span class="nc" id="L30">        List&lt;ConsequenceType&gt; consequenceTypeList = new ArrayList&lt;&gt;();</span>
<span class="nc" id="L31">        variant = inputVariant;</span>
<span class="nc" id="L32">        variantEnd = variant.getStart() + variant.getReference().length() - 1;</span>
<span class="nc" id="L33">        variantStart = variant.getStart();</span>
<span class="nc" id="L34">        boolean isIntergenic = true;</span>
<span class="nc bnc" id="L35" title="All 2 branches missed.">        for (Gene currentGene : geneList) {</span>
<span class="nc" id="L36">            gene = currentGene;</span>
<span class="nc bnc" id="L37" title="All 2 branches missed.">            for (Transcript currentTranscript : gene.getTranscripts()) {</span>
<span class="nc bnc" id="L38" title="All 6 branches missed.">                isIntergenic = isIntergenic &amp;&amp; (variantEnd &lt; currentTranscript.getStart() || variantStart &gt; currentTranscript.getEnd());</span>
<span class="nc" id="L39">                transcript = currentTranscript;</span>
<span class="nc" id="L40">                consequenceType = new ConsequenceType();</span>
<span class="nc" id="L41">                consequenceType.setGeneName(gene.getName());</span>
<span class="nc" id="L42">                consequenceType.setEnsemblGeneId(gene.getId());</span>
<span class="nc" id="L43">                consequenceType.setEnsemblTranscriptId(transcript.getId());</span>
<span class="nc" id="L44">                consequenceType.setStrand(transcript.getStrand());</span>
<span class="nc" id="L45">                consequenceType.setBiotype(transcript.getBiotype());</span>
<span class="nc bnc" id="L46" title="All 2 branches missed.">                consequenceType.setTranscriptAnnotationFlags(transcript.getAnnotationFlags() != null</span>
<span class="nc" id="L47">                        ? new ArrayList&lt;&gt;(transcript.getAnnotationFlags()) : null);</span>
<span class="nc" id="L48">                SoNames.clear();</span>

<span class="nc bnc" id="L50" title="All 2 branches missed.">                if (transcript.getStrand().equals(&quot;+&quot;)) {</span>
                    // Deletion - whole transcript removed
<span class="nc bnc" id="L52" title="All 4 branches missed.">                    if (variantStart &lt;= transcript.getStart() &amp;&amp; variantEnd &gt;= transcript.getEnd()) {</span>
<span class="nc" id="L53">                        SoNames.add(VariantAnnotationUtils.TRANSCRIPT_ABLATION);</span>
<span class="nc" id="L54">                        consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L55">                        consequenceTypeList.add(consequenceType);</span>
<span class="nc bnc" id="L56" title="All 2 branches missed.">                    } else if (regionsOverlap(transcript.getStart(), transcript.getEnd(), variantStart, variantEnd)) {</span>
<span class="nc" id="L57">                        solvePositiveTranscript(consequenceTypeList);</span>
                    } else {
<span class="nc" id="L59">                        solveTranscriptFlankingRegions(VariantAnnotationUtils.UPSTREAM_GENE_VARIANT,</span>
                                VariantAnnotationUtils.DOWNSTREAM_GENE_VARIANT);
<span class="nc bnc" id="L61" title="All 2 branches missed.">                        if (SoNames.size() &gt; 0) { // Variant does not overlap gene region, just may have upstream/downstream annotations</span>
<span class="nc" id="L62">                            consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L63">                            consequenceTypeList.add(consequenceType);</span>
                        }
                    }
                } else {
<span class="nc bnc" id="L67" title="All 4 branches missed.">                    if (variantStart &lt;= transcript.getStart() &amp;&amp; variantEnd &gt;= transcript.getEnd()) { // Deletion - whole trans. removed</span>
<span class="nc" id="L68">                        SoNames.add(VariantAnnotationUtils.TRANSCRIPT_ABLATION);</span>
<span class="nc" id="L69">                        consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L70">                        consequenceTypeList.add(consequenceType);</span>
<span class="nc bnc" id="L71" title="All 2 branches missed.">                    } else if (regionsOverlap(transcript.getStart(), transcript.getEnd(), variantStart, variantEnd)) {</span>
<span class="nc" id="L72">                        solveNegativeTranscript(consequenceTypeList);</span>
                    } else {
<span class="nc" id="L74">                        solveTranscriptFlankingRegions(VariantAnnotationUtils.DOWNSTREAM_GENE_VARIANT,</span>
                                VariantAnnotationUtils.UPSTREAM_GENE_VARIANT);
<span class="nc bnc" id="L76" title="All 2 branches missed.">                        if (SoNames.size() &gt; 0) { // Variant does not overlap gene region, just has upstream/downstream annotations</span>
<span class="nc" id="L77">                            consequenceType.setSequenceOntologyTerms(getSequenceOntologyTerms(SoNames));</span>
<span class="nc" id="L78">                            consequenceTypeList.add(consequenceType);</span>
                        }
                    }
                }
<span class="nc" id="L82">            }</span>
<span class="nc" id="L83">        }</span>

<span class="nc" id="L85">        solveIntergenic(consequenceTypeList, isIntergenic);</span>
<span class="nc" id="L86">        solveRegulatoryRegions(overlapsRegulatoryRegion, consequenceTypeList);</span>
<span class="nc" id="L87">        return consequenceTypeList;</span>
    }

    protected void solveCodingExonVariantInNegativeTranscript(boolean splicing, String transcriptSequence, int cdnaCodingStart,
                                                            int cdnaVariantStart, int cdnaVariantEnd) {
<span class="nc" id="L92">        Boolean codingAnnotationAdded = false;</span>

        // cdnaVariantStart=null if variant is intronic. cdnaCodingStart&lt;1 if cds_start_NF and phase!=0
<span class="nc bnc" id="L95" title="All 6 branches missed.">        if (cdnaVariantStart != -1 &amp;&amp; cdnaVariantStart &lt; (cdnaCodingStart + 3) &amp;&amp; (cdnaCodingStart &gt; 0</span>
<span class="nc bnc" id="L96" title="All 2 branches missed.">                || !transcript.unconfirmedStart())) {</span>
<span class="nc" id="L97">            SoNames.add(VariantAnnotationUtils.START_LOST);</span>
<span class="nc" id="L98">            codingAnnotationAdded = true;</span>
        }
<span class="nc bnc" id="L100" title="All 2 branches missed.">        if (cdnaVariantEnd != -1) {</span>
<span class="nc" id="L101">            int finalNtPhase = (transcript.getCdnaCodingEnd() - cdnaCodingStart) % 3;</span>
<span class="nc" id="L102">            Boolean stopToSolve = true;</span>
            // just checks cdnaVariantStart!=null because no splicing means cdnaVariantEnd is also != null
<span class="nc bnc" id="L104" title="All 4 branches missed.">            if (!splicing &amp;&amp; cdnaVariantStart != -1) {</span>
<span class="nc" id="L105">                codingAnnotationAdded = true;</span>
<span class="nc bnc" id="L106" title="All 2 branches missed.">                if (Math.abs(variant.getReference().length() - variant.getAlternate().length()) % 3 == 0) {</span>
<span class="nc" id="L107">                    SoNames.add(VariantAnnotationUtils.INFRAME_VARIANT);</span>
                } else {
<span class="nc" id="L109">                    SoNames.add(VariantAnnotationUtils.FRAMESHIFT_VARIANT);</span>
                }
<span class="nc" id="L111">                stopToSolve = false;  // Stop codon annotation will be solved in the line below.</span>
<span class="nc" id="L112">                solveStopCodonNegativeMNV(transcriptSequence, cdnaCodingStart, cdnaVariantStart, cdnaVariantEnd);</span>
            }
<span class="nc bnc" id="L114" title="All 2 branches missed.">            if (cdnaVariantEnd &gt;= (transcript.getCdnaCodingEnd() - finalNtPhase)) {</span>
<span class="nc bnc" id="L115" title="All 2 branches missed.">                if (finalNtPhase != 2) {</span>
<span class="nc" id="L116">                    SoNames.add(VariantAnnotationUtils.INCOMPLETE_TERMINAL_CODON_VARIANT);</span>
<span class="nc bnc" id="L117" title="All 2 branches missed.">                } else if (stopToSolve) {  // Only if stop codon annotation was not already solved in the if block above</span>
<span class="nc" id="L118">                    SoNames.add(VariantAnnotationUtils.STOP_LOST);</span>
                }
            }
        }
<span class="nc bnc" id="L122" title="All 2 branches missed.">        if (!codingAnnotationAdded) {</span>
<span class="nc" id="L123">            SoNames.add(VariantAnnotationUtils.CODING_SEQUENCE_VARIANT);</span>
        }
<span class="nc" id="L125">    }</span>

    private void solveStopCodonNegativeMNV(String transcriptSequence, int cdnaCodingStart,
                                           int cdnaVariantStart, int cdnaVariantEnd) {
<span class="nc" id="L129">        Integer variantPhaseShift1 = (cdnaVariantStart - cdnaCodingStart) % 3;</span>
<span class="nc" id="L130">        Integer variantPhaseShift2 = (cdnaVariantEnd - cdnaCodingStart) % 3;</span>
<span class="nc" id="L131">        int modifiedCodon1Start = cdnaVariantStart - variantPhaseShift1;</span>
<span class="nc" id="L132">        int modifiedCodon2Start = cdnaVariantEnd - variantPhaseShift2;</span>
<span class="nc bnc" id="L133" title="All 4 branches missed.">        if (modifiedCodon1Start &gt; 0 &amp;&amp; (modifiedCodon2Start + 2) &lt;= transcriptSequence.length()) {</span>
            /**
             * Calculate reference codon
             */
<span class="nc" id="L137">            String reverseCodon1 = new StringBuilder(transcriptSequence.substring(transcriptSequence.length() - modifiedCodon1Start - 2,</span>
                    // Rigth limit of the substring sums +1 because substring does not include that position
<span class="nc" id="L139">                    transcriptSequence.length() - modifiedCodon1Start + 1)).reverse().toString();</span>
<span class="nc" id="L140">            String reverseCodon2 = new StringBuilder(transcriptSequence.substring(transcriptSequence.length() - modifiedCodon2Start - 2,</span>
                    // Rigth limit of the substring sums +1 because substring does not include that position
<span class="nc" id="L142">                    transcriptSequence.length() - modifiedCodon2Start + 1)).reverse().toString();</span>
<span class="nc" id="L143">            String reverseTranscriptSequence = new StringBuilder(</span>
<span class="nc bnc" id="L144" title="All 2 branches missed.">                    transcriptSequence.substring(((transcriptSequence.length() - cdnaVariantEnd) &gt; 2)</span>
<span class="nc" id="L145">                                    ? (transcriptSequence.length() - cdnaVariantEnd - 3)</span>
                                    : 0,  // Be careful reaching the end of the transcript sequence
                            // Rigth limit of the substring -2 because substring does not include that position
<span class="nc" id="L148">                            transcriptSequence.length() - cdnaVariantEnd)).reverse().toString();</span>
<span class="nc" id="L149">            char[] referenceCodon1Array = reverseCodon1.toCharArray();</span>
<span class="nc" id="L150">            referenceCodon1Array[0] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon1Array[0]);</span>
<span class="nc" id="L151">            referenceCodon1Array[1] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon1Array[1]);</span>
<span class="nc" id="L152">            referenceCodon1Array[2] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon1Array[2]);</span>
<span class="nc" id="L153">            String referenceCodon1 = String.valueOf(referenceCodon1Array);</span>
<span class="nc" id="L154">            char[] referenceCodon2Array = reverseCodon2.toCharArray();</span>
<span class="nc" id="L155">            referenceCodon2Array[0] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon2Array[0]);</span>
<span class="nc" id="L156">            referenceCodon2Array[1] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon2Array[1]);</span>
<span class="nc" id="L157">            referenceCodon2Array[2] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(referenceCodon2Array[2]);</span>
<span class="nc" id="L158">            String referenceCodon2 = String.valueOf(referenceCodon2Array);</span>


            // Char array to contain the upper/lower-case formatted string for the codon change, e.g. aGT/ATG
<span class="nc" id="L162">            char[] formattedReferenceCodon1Array = String.valueOf(referenceCodon1Array).toLowerCase().toCharArray();</span>


            /**
             * Calculate codons generated by the insertion
             */
<span class="nc" id="L168">            int i = 0;</span>
<span class="nc" id="L169">            int reverseTranscriptSequencePosition = 0;</span>
<span class="nc" id="L170">            char[] altArray = (new StringBuilder(variant.getAlternate()).reverse().toString()).toCharArray();</span>
            int modifiedCodonPosition;
<span class="nc" id="L172">            int modifiedCodonPositionStart = variantPhaseShift1;</span>
<span class="nc" id="L173">            char[] formattedModifiedCodonArray = referenceCodon1.toLowerCase().toCharArray();</span>
<span class="nc" id="L174">            char[] modifiedCodonArray = referenceCodon1Array.clone();</span>
<span class="nc" id="L175">            boolean useMitochondrialCode = variant.getChromosome().equals(&quot;MT&quot;);</span>
<span class="nc" id="L176">            boolean firstCodon = true;</span>

            // Solving the stop codon is more less equivalent to dealing with a deletion followed by an insertion
            do {
<span class="nc" id="L180">                for (modifiedCodonPosition = modifiedCodonPositionStart;</span>
                    // Paste alternative nt in the corresponding codon position
<span class="nc bnc" id="L182" title="All 4 branches missed.">                     (modifiedCodonPosition &lt; 3 &amp;&amp; i &lt; variant.getAlternate().length()); modifiedCodonPosition++) {</span>
<span class="nc" id="L183">                    modifiedCodonArray[modifiedCodonPosition] = VariantAnnotationUtils.COMPLEMENTARY_NT.get(altArray[i]);</span>

                    // Edit modified nt to make it upper-case in the formatted strings
<span class="nc" id="L186">                    formattedReferenceCodon1Array[modifiedCodonPosition]</span>
<span class="nc" id="L187">                            = Character.toUpperCase(formattedReferenceCodon1Array[modifiedCodonPosition]);</span>
<span class="nc" id="L188">                    formattedModifiedCodonArray[modifiedCodonPosition]</span>
<span class="nc" id="L189">                            = Character.toUpperCase(modifiedCodonArray[modifiedCodonPosition]);</span>

<span class="nc" id="L191">                    i++;</span>
                }
<span class="nc" id="L193">                reverseTranscriptSequencePosition = updateNegativeInsertionCodonArrays(reverseTranscriptSequence,</span>
                        formattedReferenceCodon1Array, reverseTranscriptSequencePosition, modifiedCodonPosition,
                        formattedModifiedCodonArray, modifiedCodonArray);

                // Set codon str, protein ref and protein alt ONLY for the first codon mofified by the insertion
<span class="nc" id="L198">                firstCodon = setInsertionAlleleAminoacidChange(referenceCodon1, modifiedCodonArray,</span>
                        formattedReferenceCodon1Array, formattedModifiedCodonArray, useMitochondrialCode, firstCodon);

<span class="nc" id="L201">                decideStopCodonModificationAnnotation(SoNames,</span>
<span class="nc bnc" id="L202" title="All 2 branches missed.">                        VariantAnnotationUtils.isStopCodon(useMitochondrialCode, referenceCodon2)</span>
<span class="nc" id="L203">                                ? referenceCodon2 : referenceCodon1, String.valueOf(modifiedCodonArray),</span>
                        useMitochondrialCode);

<span class="nc" id="L206">                modifiedCodonPositionStart = 0;  // Reset the position where the next modified codon must be started to be filled</span>
<span class="nc bnc" id="L207" title="All 2 branches missed.">            } while (i &lt; variant.getAlternate().length());  // All posible new codons generated by the inserted sequence must be checked</span>

        }
<span class="nc" id="L210">    }</span>

    protected void solveCodingExonVariantInPositiveTranscript(boolean splicing, String transcriptSequence, int cdnaCodingStart,
                                                            int cdnaVariantStart, int cdnaVariantEnd) {
        // This will indicate wether it is needed to add the &quot;coding_sequence_variant&quot; annotation or not
<span class="nc" id="L215">        boolean codingAnnotationAdded = false;</span>

        // cdnaVariantStart=null if variant is intronic. cdnaCodingStart&lt;1 if cds_start_NF and phase!=0
<span class="nc bnc" id="L218" title="All 8 branches missed.">        if (cdnaVariantStart != -1 &amp;&amp; cdnaVariantStart &lt; (cdnaCodingStart + 3) &amp;&amp; (cdnaCodingStart &gt; 0 || !transcript.unconfirmedStart())) {</span>
<span class="nc" id="L219">            SoNames.add(VariantAnnotationUtils.START_LOST);</span>
<span class="nc" id="L220">            codingAnnotationAdded = true;</span>
        }
<span class="nc bnc" id="L222" title="All 2 branches missed.">        if (cdnaVariantEnd != -1) {</span>
<span class="nc" id="L223">            int finalNtPhase = (transcript.getCdnaCodingEnd() - cdnaCodingStart) % 3;</span>
<span class="nc" id="L224">            Boolean stopToSolve = true;</span>
            // just checks cdnaVariantStart!=null because no splicing means cdnaVariantEnd is also != null
<span class="nc bnc" id="L226" title="All 4 branches missed.">            if (!splicing &amp;&amp; cdnaVariantStart != -1) {</span>
<span class="nc" id="L227">                codingAnnotationAdded = true;</span>
<span class="nc bnc" id="L228" title="All 2 branches missed.">                if (Math.abs(variant.getReference().length() - variant.getAlternate().length()) % 3 == 0) {</span>
<span class="nc" id="L229">                    SoNames.add(VariantAnnotationUtils.INFRAME_VARIANT);</span>
                } else {
<span class="nc" id="L231">                    SoNames.add(VariantAnnotationUtils.FRAMESHIFT_VARIANT);</span>
                }
<span class="nc" id="L233">                stopToSolve = false;  // Stop codon annotation will be solved in the line below.</span>
<span class="nc" id="L234">                solveStopCodonPositiveMNV(transcriptSequence, cdnaCodingStart, cdnaVariantStart, cdnaVariantEnd);</span>
            }
<span class="nc bnc" id="L236" title="All 2 branches missed.">            if (cdnaVariantEnd &gt;= (transcript.getCdnaCodingEnd() - finalNtPhase)) {</span>
<span class="nc bnc" id="L237" title="All 2 branches missed.">                if (finalNtPhase != 2) {</span>
//                if (transcript.unconfirmedEnd() &amp;&amp; (finalNtPhase != 2)) {
<span class="nc" id="L239">                    SoNames.add(VariantAnnotationUtils.INCOMPLETE_TERMINAL_CODON_VARIANT);</span>
<span class="nc bnc" id="L240" title="All 2 branches missed.">                } else if (stopToSolve) {  // Only if stop codon annotation was not already solved in the if block above</span>
<span class="nc" id="L241">                    SoNames.add(VariantAnnotationUtils.STOP_LOST);</span>
                }
            }
        }
<span class="nc bnc" id="L245" title="All 2 branches missed.">        if (!codingAnnotationAdded) {</span>
<span class="nc" id="L246">            SoNames.add(VariantAnnotationUtils.CODING_SEQUENCE_VARIANT);</span>
        }
<span class="nc" id="L248">    }</span>

    private void solveStopCodonPositiveMNV(String transcriptSequence, int cdnaCodingStart, int cdnaVariantStart,
                                           int cdnaVariantEnd) {
<span class="nc" id="L252">        Integer variantPhaseShift1 = (cdnaVariantStart - cdnaCodingStart) % 3;</span>
<span class="nc" id="L253">        Integer variantPhaseShift2 = (cdnaVariantEnd - cdnaCodingStart) % 3;</span>
<span class="nc" id="L254">        int modifiedCodon1Start = cdnaVariantStart - variantPhaseShift1;</span>
<span class="nc" id="L255">        int modifiedCodon2Start = cdnaVariantEnd - variantPhaseShift2;</span>
<span class="nc bnc" id="L256" title="All 4 branches missed.">        if (modifiedCodon1Start &gt; 0 &amp;&amp; (modifiedCodon2Start + 2) &lt;= transcriptSequence.length()) {</span>
            /**
             * Calculate reference codons
             */
            // -1 and +2 because of base 0 String indexing
<span class="nc" id="L261">            String referenceCodon1 = transcriptSequence.substring(modifiedCodon1Start - 1, modifiedCodon1Start + 2);</span>
<span class="nc" id="L262">            String referenceCodon2 = transcriptSequence.substring(modifiedCodon2Start - 1, modifiedCodon2Start + 2);</span>
<span class="nc" id="L263">            int i = cdnaVariantEnd;  // Position (0 based index) in transcriptSequence of the first nt after the deletion</span>
            int codonPosition;

            // Char array to contain the upper/lower-case formatted strings for the codon change, e.g. aGT/ATG
<span class="nc" id="L267">            char[] formattedReferenceCodon1Array = referenceCodon1.toLowerCase().toCharArray();</span>

            /**
             * Calculate codons generated by the insertion
             */
<span class="nc" id="L272">            char[] modifiedCodonArray = referenceCodon1.toCharArray();</span>
            // indexing over transcriptSequence is 0 based, cdnaVariantEnd points to the first position that remains in
            // the allele
<span class="nc" id="L275">            int transcriptSequencePosition = cdnaVariantEnd;</span>
//            int transcriptSequencePosition = cdnaVariantStart;
            int modifiedCodonPosition;
<span class="nc" id="L278">            int modifiedCodonPositionStart = variantPhaseShift1;</span>

            // Char arrays to contain the upper/lower-case formatted strings for the codon change, e.g. aGT/ATG
//            char[] formattedReferenceCodonArray = referenceCodon.toLowerCase().toCharArray();
<span class="nc" id="L282">            char[] formattedModifiedCodonArray = referenceCodon1.toLowerCase().toCharArray();</span>
<span class="nc" id="L283">            boolean useMitochondrialCode = variant.getChromosome().equals(&quot;MT&quot;);</span>
<span class="nc" id="L284">            boolean firstCodon = true;</span>
<span class="nc" id="L285">            i = 0;</span>

            // Solving the stop codon is more less equivalent to dealing with a deletion followed by an insertion
            do {
<span class="nc" id="L289">                for (modifiedCodonPosition = modifiedCodonPositionStart;</span>
                    // Paste alternative nt in the corresponding codon position
<span class="nc bnc" id="L291" title="All 4 branches missed.">                     (modifiedCodonPosition &lt; 3 &amp;&amp; i &lt; variant.getAlternate().length()); modifiedCodonPosition++) {</span>
<span class="nc" id="L292">                    modifiedCodonArray[modifiedCodonPosition] = variant.getAlternate().toCharArray()[i];</span>

                    // Edit modified nt to make it upper-case in the formatted strings
<span class="nc" id="L295">                    formattedReferenceCodon1Array[modifiedCodonPosition]</span>
<span class="nc" id="L296">                            = Character.toUpperCase(formattedReferenceCodon1Array[modifiedCodonPosition]);</span>
<span class="nc" id="L297">                    formattedModifiedCodonArray[modifiedCodonPosition]</span>
<span class="nc" id="L298">                            = Character.toUpperCase(variant.getAlternate().toCharArray()[i]);</span>

<span class="nc" id="L300">                    i++;</span>
                }
<span class="nc" id="L302">                transcriptSequencePosition = updatePositiveInsertionCodonArrays(transcriptSequence, modifiedCodonArray,</span>
                        transcriptSequencePosition, modifiedCodonPosition, formattedReferenceCodon1Array,
                        formattedModifiedCodonArray);

<span class="nc" id="L306">                firstCodon = setInsertionAlleleAminoacidChange(referenceCodon1, modifiedCodonArray,</span>
                        formattedReferenceCodon1Array, formattedModifiedCodonArray, useMitochondrialCode, firstCodon);

<span class="nc" id="L309">                decideStopCodonModificationAnnotation(SoNames,</span>
<span class="nc bnc" id="L310" title="All 2 branches missed.">                        VariantAnnotationUtils.isStopCodon(useMitochondrialCode, referenceCodon2)</span>
<span class="nc" id="L311">                                ? referenceCodon2 : referenceCodon1, String.valueOf(modifiedCodonArray),</span>
                        useMitochondrialCode);
<span class="nc" id="L313">                modifiedCodonPositionStart = 0;  // Reset the position where the next modified codon must be started to be filled</span>
<span class="nc bnc" id="L314" title="All 2 branches missed.">            } while (i &lt; variant.getAlternate().length());  // All posible new codons generated by the inserted sequence must be checked</span>
        }
<span class="nc" id="L316">    }</span>

    protected void solveJunction(Integer spliceSite1, Integer spliceSite2, String leftSpliceSiteTag,
                               String rightSpliceSiteTag, boolean[] junctionSolution) {

<span class="nc" id="L321">        junctionSolution[0] = false;  // Is splicing variant in non-coding region</span>
<span class="nc" id="L322">        junctionSolution[1] = false;  // Variant is intronic and both ends fall within the intron</span>

        // Variant overlaps the rest of intronic region (splice region within the intron and/or rest of intron)
<span class="nc bnc" id="L325" title="All 2 branches missed.">        if (regionsOverlap(spliceSite1 + 2, spliceSite2 - 2, variantStart, variantEnd)) {</span>
<span class="nc" id="L326">            SoNames.add(VariantAnnotationUtils.INTRON_VARIANT);</span>
        }
<span class="nc bnc" id="L328" title="All 4 branches missed.">        if (variantStart &gt;= spliceSite1 &amp;&amp; variantEnd &lt;= spliceSite2) {</span>
<span class="nc" id="L329">            junctionSolution[1] = true;  // variant start &amp; end fall within the intron</span>
        }

<span class="nc bnc" id="L332" title="All 2 branches missed.">        if (regionsOverlap(spliceSite1, spliceSite1 + 1, variantStart, variantEnd)) {  // Variant donor/acceptor</span>
<span class="nc" id="L333">            SoNames.add(leftSpliceSiteTag);  // donor/acceptor depending on transcript strand</span>
            // BE CAREFUL: there are introns shorter than 7nts, and even just 1nt long!! (22:36587846)
<span class="nc bnc" id="L335" title="All 4 branches missed.">            junctionSolution[0] = (variantStart &lt;= spliceSite2 || variantEnd &lt;= spliceSite2);</span>
<span class="nc bnc" id="L336" title="All 2 branches missed.">        } else if (regionsOverlap(spliceSite1 + 2, spliceSite1 + 7, variantStart, variantEnd)) {</span>
<span class="nc" id="L337">            SoNames.add(VariantAnnotationUtils.SPLICE_REGION_VARIANT);</span>
<span class="nc bnc" id="L338" title="All 2 branches missed.">        } else if (regionsOverlap(spliceSite1 - 3, spliceSite1 - 1, variantStart, variantEnd)) {</span>
<span class="nc" id="L339">            SoNames.add(VariantAnnotationUtils.SPLICE_REGION_VARIANT);</span>
        }

<span class="nc bnc" id="L342" title="All 2 branches missed.">        if (regionsOverlap(spliceSite2 - 1, spliceSite2, variantStart, variantEnd)) {  // Variant donor/acceptor</span>
<span class="nc" id="L343">            SoNames.add(rightSpliceSiteTag);  // donor/acceptor depending on transcript strand</span>
            // BE CAREFUL: there are introns shorter than 7nts, and even just 1nt long!! (22:36587846)
<span class="nc bnc" id="L345" title="All 4 branches missed.">            junctionSolution[0] = (spliceSite1 &lt;= variantStart || spliceSite1 &lt;= variantEnd);</span>
<span class="nc bnc" id="L346" title="All 2 branches missed.">        } else if (regionsOverlap(spliceSite2 - 7, spliceSite2 - 2, variantStart, variantEnd)) {</span>
<span class="nc" id="L347">            SoNames.add(VariantAnnotationUtils.SPLICE_REGION_VARIANT);</span>
            // BE CAREFUL: there are introns shorter than 7nts, and even just 1nt long!! (22:36587846)
<span class="nc bnc" id="L349" title="All 4 branches missed.">            junctionSolution[0] = (spliceSite1 &lt;= variantStart || spliceSite1 &lt;= variantEnd);</span>
<span class="nc bnc" id="L350" title="All 2 branches missed.">        } else if (regionsOverlap(spliceSite2 + 1, spliceSite2 + 3, variantStart, variantEnd)) {</span>
<span class="nc" id="L351">            SoNames.add(VariantAnnotationUtils.SPLICE_REGION_VARIANT);</span>
        }
<span class="nc" id="L353">    }</span>

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>