<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>GeneParserUtils.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-app</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.app.transform</a> &gt; <span class="el_source">GeneParserUtils.java</span></div><h1>GeneParserUtils.java</h1><pre class="source lang-java linenums">/*
 * Copyright 2015 OpenCB
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.opencb.cellbase.app.transform;

import org.opencb.biodata.formats.feature.gff.Gff2;
import org.opencb.biodata.formats.feature.gff.io.Gff2Reader;
import org.opencb.biodata.formats.io.FileFormatException;
import org.opencb.biodata.models.core.MiRNAGene;
import org.opencb.biodata.models.core.Xref;
import org.opencb.biodata.models.variant.avro.Expression;
import org.opencb.biodata.models.variant.avro.ExpressionCall;
import org.opencb.biodata.models.variant.avro.GeneDrugInteraction;
import org.opencb.biodata.models.variant.avro.GeneTraitAssociation;
import org.opencb.commons.utils.FileUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

/**
 * Created by imedina on 12/11/15.
 */
<span class="nc" id="L42">public class GeneParserUtils {</span>

<span class="fc" id="L44">    private static Logger logger = LoggerFactory.getLogger(GeneParserUtils.class);</span>

    public static Map&lt;String, SortedSet&lt;Gff2&gt;&gt; getTfbsMap(Path tfbsFile) throws IOException, NoSuchMethodException, FileFormatException {
<span class="fc" id="L47">        Map&lt;String, SortedSet&lt;Gff2&gt;&gt; tfbsMap = new HashMap&lt;&gt;();</span>

<span class="pc bpc" id="L49" title="3 of 8 branches missed.">        if (tfbsFile != null &amp;&amp; Files.exists(tfbsFile) &amp;&amp; !Files.isDirectory(tfbsFile) &amp;&amp; Files.size(tfbsFile) &gt; 0) {</span>
<span class="fc" id="L50">            Gff2Reader motifsFeatureReader = new Gff2Reader(tfbsFile);</span>
            Gff2 tfbsMotifFeature;
<span class="fc bfc" id="L52" title="All 2 branches covered.">            while ((tfbsMotifFeature = motifsFeatureReader.read()) != null) {</span>
<span class="fc" id="L53">                String chromosome = tfbsMotifFeature.getSequenceName().replaceFirst(&quot;chr&quot;, &quot;&quot;);</span>
<span class="fc" id="L54">                SortedSet&lt;Gff2&gt; chromosomeTfbsSet = tfbsMap.get(chromosome);</span>
<span class="fc bfc" id="L55" title="All 2 branches covered.">                if (chromosomeTfbsSet == null) {</span>
<span class="fc" id="L56">                    chromosomeTfbsSet = new TreeSet&lt;&gt;((Comparator&lt;Gff2&gt;) (feature1, feature2) -&gt; {</span>
                        // TODO: maybe this should be in TranscriptTfbs class, and equals method should be overriden too
<span class="fc bfc" id="L58" title="All 2 branches covered.">                        if (feature1.getStart() != feature2.getStart()) {</span>
<span class="fc" id="L59">                            return feature1.getStart() - feature2.getStart();</span>
                        } else {
<span class="fc" id="L61">                            return feature1.getAttribute().compareTo(feature2.getAttribute());</span>
                        }
                    });
<span class="fc" id="L64">                    tfbsMap.put(chromosome, chromosomeTfbsSet);</span>
                }
<span class="fc" id="L66">                chromosomeTfbsSet.add(tfbsMotifFeature);</span>
<span class="fc" id="L67">            }</span>
<span class="fc" id="L68">            motifsFeatureReader.close();</span>
        }
<span class="fc" id="L70">        return tfbsMap;</span>
    }

    public static Map&lt;String, MiRNAGene&gt; getmiRNAGeneMap(Path mirnaGeneFile) throws IOException {
<span class="fc" id="L74">        Map&lt;String, MiRNAGene&gt; mirnaGeneMap = new HashMap&lt;&gt;();</span>

<span class="pc bpc" id="L76" title="4 of 6 branches missed.">        if (mirnaGeneFile != null &amp;&amp; Files.exists(mirnaGeneFile) &amp;&amp; !Files.isDirectory(mirnaGeneFile)</span>
<span class="nc bnc" id="L77" title="All 2 branches missed.">                &amp;&amp; Files.size(mirnaGeneFile) &gt; 0) {</span>
<span class="nc" id="L78">            logger.info(&quot;Loading miRNA data ...&quot;);</span>
<span class="nc" id="L79">            BufferedReader br = Files.newBufferedReader(mirnaGeneFile, Charset.defaultCharset());</span>

            String line;
            String[] fields, mirnaMatures, mirnaMaturesFields;
            List&lt;String&gt; aliases;
            MiRNAGene miRNAGene;
<span class="nc bnc" id="L85" title="All 2 branches missed.">            while ((line = br.readLine()) != null) {</span>
<span class="nc" id="L86">                fields = line.split(&quot;\t&quot;);</span>

                // First, read aliases of miRNA, field #5
<span class="nc" id="L89">                aliases = new ArrayList&lt;&gt;();</span>
<span class="nc bnc" id="L90" title="All 2 branches missed.">                for (String alias : fields[5].split(&quot;,&quot;)) {</span>
<span class="nc" id="L91">                    aliases.add(alias);</span>
                }

<span class="nc" id="L94">                miRNAGene = new MiRNAGene(fields[1], fields[2], fields[3], fields[4], aliases, new ArrayList&lt;&gt;());</span>

                // Second, read the miRNA matures, field #6
<span class="nc" id="L97">                mirnaMatures = fields[6].split(&quot;,&quot;);</span>
<span class="nc bnc" id="L98" title="All 2 branches missed.">                for (String s : mirnaMatures) {</span>
<span class="nc" id="L99">                    mirnaMaturesFields = s.split(&quot;\\|&quot;);</span>
<span class="nc" id="L100">                    int cdnaStart = fields[4].indexOf(mirnaMaturesFields[2]) + 1;</span>
<span class="nc" id="L101">                    int cdnaEnd = cdnaStart + mirnaMaturesFields[2].length() - 1;</span>
                    // Save directly into MiRNAGene object.
<span class="nc" id="L103">                    miRNAGene.addMiRNAMature(mirnaMaturesFields[0], mirnaMaturesFields[1], mirnaMaturesFields[2], cdnaStart, cdnaEnd);</span>
                }

                // Add object to Map&lt;EnsemblID, MiRNAGene&gt;
<span class="nc" id="L107">                mirnaGeneMap.put(fields[0], miRNAGene);</span>
            }
<span class="nc" id="L109">            br.close();</span>
<span class="nc" id="L110">        } else {</span>
<span class="fc" id="L111">            logger.warn(&quot;Mirna file '{}' not found&quot;, mirnaGeneFile);</span>
<span class="fc" id="L112">            logger.warn(&quot;Mirna data not loaded&quot;);</span>
        }
<span class="fc" id="L114">        return mirnaGeneMap;</span>
    }


    public static Map&lt;String, ArrayList&lt;Xref&gt;&gt; getXrefMap(Path xrefsFile, Path uniprotIdMappingFile) throws IOException {
<span class="fc" id="L119">        Map&lt;String, ArrayList&lt;Xref&gt;&gt; xrefMap = new HashMap&lt;&gt;();</span>
<span class="fc" id="L120">        logger.info(&quot;Loading xref data...&quot;);</span>
        String[] fields;
<span class="pc bpc" id="L122" title="4 of 6 branches missed.">        if (xrefsFile != null &amp;&amp; Files.exists(xrefsFile) &amp;&amp; Files.size(xrefsFile) &gt; 0) {</span>
<span class="nc" id="L123">            List&lt;String&gt; lines = Files.readAllLines(xrefsFile, Charset.forName(&quot;ISO-8859-1&quot;));</span>
//            List&lt;String&gt; lines = Files.readAllLines(xrefsFile, Charset.defaultCharset());
<span class="nc bnc" id="L125" title="All 2 branches missed.">            for (String line : lines) {</span>
<span class="nc" id="L126">                fields = line.split(&quot;\t&quot;, -1);</span>
<span class="nc bnc" id="L127" title="All 2 branches missed.">                if (fields.length &gt;= 4) {</span>
<span class="nc bnc" id="L128" title="All 2 branches missed.">                    if (!xrefMap.containsKey(fields[0])) {</span>
<span class="nc" id="L129">                        xrefMap.put(fields[0], new ArrayList&lt;&gt;());</span>
                    }
<span class="nc" id="L131">                    xrefMap.get(fields[0]).add(new Xref(fields[1], fields[2], fields[3]));</span>
                }
<span class="nc" id="L133">            }</span>
<span class="nc" id="L134">        } else {</span>
<span class="fc" id="L135">            logger.warn(&quot;Xrefs file &quot; + xrefsFile + &quot; not found&quot;);</span>
<span class="fc" id="L136">            logger.warn(&quot;Xref data not loaded&quot;);</span>
        }

<span class="fc" id="L139">        logger.info(&quot;Loading protein mapping into xref data...&quot;);</span>
<span class="pc bpc" id="L140" title="4 of 6 branches missed.">        if (uniprotIdMappingFile != null &amp;&amp; Files.exists(uniprotIdMappingFile) &amp;&amp; Files.size(uniprotIdMappingFile) &gt; 0) {</span>
<span class="nc" id="L141">            BufferedReader br = FileUtils.newBufferedReader(uniprotIdMappingFile);</span>
            String line;
<span class="nc bnc" id="L143" title="All 2 branches missed.">            while ((line = br.readLine()) != null) {</span>
<span class="nc" id="L144">                fields = line.split(&quot;\t&quot;, -1);</span>
<span class="nc bnc" id="L145" title="All 4 branches missed.">                if (fields.length &gt;= 19 &amp;&amp; fields[19].startsWith(&quot;ENST&quot;)) {</span>
<span class="nc" id="L146">                    String[] transcripts = fields[19].split(&quot;; &quot;);</span>
<span class="nc bnc" id="L147" title="All 2 branches missed.">                    for (String transcript : transcripts) {</span>
<span class="nc bnc" id="L148" title="All 2 branches missed.">                        if (!xrefMap.containsKey(transcript)) {</span>
<span class="nc" id="L149">                            xrefMap.put(transcript, new ArrayList&lt;Xref&gt;());</span>
                        }
<span class="nc" id="L151">                        xrefMap.get(transcript).add(new Xref(fields[0], &quot;uniprotkb_acc&quot;, &quot;UniProtKB ACC&quot;));</span>
<span class="nc" id="L152">                        xrefMap.get(transcript).add(new Xref(fields[1], &quot;uniprotkb_id&quot;, &quot;UniProtKB ID&quot;));</span>
                    }
<span class="nc" id="L154">                }</span>
            }
<span class="nc" id="L156">            br.close();</span>
<span class="nc" id="L157">        } else {</span>
<span class="fc" id="L158">            logger.warn(&quot;Uniprot if mapping file &quot; + uniprotIdMappingFile + &quot; not found&quot;);</span>
<span class="fc" id="L159">            logger.warn(&quot;Protein mapping into xref data not loaded&quot;);</span>
        }

<span class="fc" id="L162">        return xrefMap;</span>
    }

    public static Map&lt;String, List&lt;GeneDrugInteraction&gt;&gt; getGeneDrugMap(Path geneDrugFile) throws IOException {
<span class="fc" id="L166">        Map&lt;String, List&lt;GeneDrugInteraction&gt;&gt; geneDrugMap = new HashMap&lt;&gt;();</span>
<span class="pc bpc" id="L167" title="4 of 6 branches missed.">        if (geneDrugFile != null &amp;&amp; Files.exists(geneDrugFile) &amp;&amp; Files.size(geneDrugFile) &gt; 0) {</span>
<span class="nc" id="L168">            logger.info(&quot;Loading gene-drug interaction data from '{}'&quot;, geneDrugFile);</span>
<span class="nc" id="L169">            BufferedReader br = FileUtils.newBufferedReader(geneDrugFile);</span>

            // Skip header
<span class="nc" id="L172">            br.readLine();</span>

<span class="nc" id="L174">            int lineCounter = 1;</span>
            String line;
<span class="nc bnc" id="L176" title="All 2 branches missed.">            while ((line = br.readLine()) != null) {</span>
<span class="nc" id="L177">                String[] parts = line.split(&quot;\t&quot;);</span>
<span class="nc" id="L178">                addValueToMapElement(geneDrugMap, parts[0], new GeneDrugInteraction(parts[0], parts[4], &quot;dgidb&quot;, parts[2], parts[3]));</span>
<span class="nc" id="L179">                lineCounter++;</span>
<span class="nc" id="L180">            }</span>

<span class="nc" id="L182">            br.close();</span>
<span class="nc" id="L183">        } else {</span>
<span class="fc" id="L184">            logger.warn(&quot;Gene drug file &quot; + geneDrugFile + &quot; not found&quot;);</span>
<span class="fc" id="L185">            logger.warn(&quot;Ignoring &quot; + geneDrugFile);</span>
        }

<span class="fc" id="L188">        return geneDrugMap;</span>
    }

    public static Map&lt;String, List&lt;Expression&gt;&gt; getGeneExpressionMap(String species, Path geneExpressionFile) throws IOException {
<span class="fc" id="L192">        Map&lt;String, List&lt;Expression&gt;&gt; geneExpressionMap = new HashMap&lt;&gt;();</span>

<span class="pc bpc" id="L194" title="6 of 8 branches missed.">        if (geneExpressionFile != null &amp;&amp; Files.exists(geneExpressionFile) &amp;&amp; Files.size(geneExpressionFile) &gt; 0</span>
                &amp;&amp; species != null) {
<span class="nc" id="L196">            logger.info(&quot;Loading gene expression data from '{}'&quot;, geneExpressionFile);</span>
<span class="nc" id="L197">            BufferedReader br = FileUtils.newBufferedReader(geneExpressionFile);</span>

            // Skip header. Column name line does not start with # so the last line read by this while will be this one
<span class="nc" id="L200">            int lineCounter = 0;</span>
            String line;
<span class="nc bnc" id="L202" title="All 2 branches missed.">            while (((line = br.readLine()) != null)) {  //  &amp;&amp; (line.startsWith(&quot;#&quot;))</span>
<span class="nc bnc" id="L203" title="All 2 branches missed.">                if (line.startsWith(&quot;#&quot;)) {</span>
<span class="nc" id="L204">                    lineCounter++;</span>
                } else {
                    break;
                }
            }

<span class="nc bnc" id="L210" title="All 2 branches missed.">            while ((line = br.readLine()) != null) {</span>
<span class="nc" id="L211">                String[] parts = line.split(&quot;\t&quot;);</span>
<span class="nc bnc" id="L212" title="All 2 branches missed.">                if (species.equals(parts[2])) {</span>
<span class="nc bnc" id="L213" title="All 2 branches missed.">                    if (parts[7].equals(&quot;UP&quot;)) {</span>
<span class="nc" id="L214">                        addValueToMapElement(geneExpressionMap, parts[1], new Expression(parts[1], null, parts[3],</span>
<span class="nc" id="L215">                                parts[4], parts[5], parts[6], ExpressionCall.UP, Float.valueOf(parts[8])));</span>
<span class="nc bnc" id="L216" title="All 2 branches missed.">                    } else if (parts[7].equals(&quot;DOWN&quot;)) {</span>
<span class="nc" id="L217">                        addValueToMapElement(geneExpressionMap, parts[1], new Expression(parts[1], null, parts[3],</span>
<span class="nc" id="L218">                                parts[4], parts[5], parts[6], ExpressionCall.DOWN, Float.valueOf(parts[8])));</span>
                    } else {
<span class="nc" id="L220">                        logger.warn(&quot;Expression tags found different from UP/DOWN at line {}. Entry omitted. &quot;, lineCounter);</span>
                    }
                }
<span class="nc" id="L223">                lineCounter++;</span>
<span class="nc" id="L224">            }</span>

<span class="nc" id="L226">            br.close();</span>
<span class="nc" id="L227">        } else {</span>
<span class="fc" id="L228">            logger.warn(&quot;Parameters are not correct&quot;);</span>
        }

<span class="fc" id="L231">        return geneExpressionMap;</span>
    }

    private static &lt;T&gt; void addValueToMapElement(Map&lt;String, List&lt;T&gt;&gt; map, String key, T value) {
<span class="nc bnc" id="L235" title="All 2 branches missed.">        if (map.containsKey(key)) {</span>
<span class="nc" id="L236">            map.get(key).add(value);</span>
        } else {
<span class="nc" id="L238">            List&lt;T&gt; expressionValueList = new ArrayList&lt;&gt;();</span>
<span class="nc" id="L239">            expressionValueList.add(value);</span>
<span class="nc" id="L240">            map.put(key, expressionValueList);</span>
        }

<span class="nc" id="L243">    }</span>


    public static Map&lt;String, List&lt;GeneTraitAssociation&gt;&gt; getGeneDiseaseAssociationMap(Path hpoFilePath, Path disgenetFilePath)
            throws IOException {
<span class="fc" id="L248">        Map&lt;String, List&lt;GeneTraitAssociation&gt;&gt; geneDiseaseAssociationMap = new HashMap&lt;&gt;(50000);</span>

        String[] fields;
        String line;
<span class="pc bpc" id="L252" title="4 of 6 branches missed.">        if (hpoFilePath != null &amp;&amp; hpoFilePath.toFile().exists() &amp;&amp; Files.size(hpoFilePath) &gt; 0) {</span>
<span class="nc" id="L253">            BufferedReader bufferedReader = FileUtils.newBufferedReader(hpoFilePath);</span>
            // skip first header line
<span class="nc" id="L255">            bufferedReader.readLine();</span>
<span class="nc bnc" id="L256" title="All 2 branches missed.">            while ((line = bufferedReader.readLine()) != null) {</span>
<span class="nc" id="L257">                fields = line.split(&quot;\t&quot;);</span>
<span class="nc" id="L258">                GeneTraitAssociation disease =</span>
<span class="nc" id="L259">                        new GeneTraitAssociation(fields[0], fields[4], fields[3], 0f, 0, new ArrayList&lt;&gt;(), new ArrayList&lt;&gt;(), &quot;hpo&quot;);</span>
<span class="nc" id="L260">                addValueToMapElement(geneDiseaseAssociationMap, fields[1], disease);</span>
<span class="nc" id="L261">            }</span>
<span class="nc" id="L262">            bufferedReader.close();</span>
        }

<span class="pc bpc" id="L265" title="4 of 6 branches missed.">        if (disgenetFilePath != null &amp;&amp; disgenetFilePath.toFile().exists() &amp;&amp; Files.size(disgenetFilePath) &gt; 0) {</span>
<span class="nc" id="L266">            BufferedReader bufferedReader = FileUtils.newBufferedReader(disgenetFilePath);</span>
            // skip first header line
<span class="nc" id="L268">            bufferedReader.readLine();</span>
<span class="nc bnc" id="L269" title="All 2 branches missed.">            while ((line = bufferedReader.readLine()) != null) {</span>
<span class="nc" id="L270">                fields = line.split(&quot;\t&quot;);</span>
<span class="nc" id="L271">                GeneTraitAssociation disease = new GeneTraitAssociation(fields[3], fields[4], &quot;&quot;, Float.parseFloat(fields[5]),</span>
<span class="nc" id="L272">                        Integer.parseInt(fields[6]), Arrays.asList(fields[7]), Arrays.asList(fields[8].split(&quot;, &quot;)), &quot;disgenet&quot;);</span>
<span class="nc" id="L273">                addValueToMapElement(geneDiseaseAssociationMap, fields[1], disease);</span>
<span class="nc" id="L274">            }</span>
<span class="nc" id="L275">            bufferedReader.close();</span>
        }

<span class="fc" id="L278">        return geneDiseaseAssociationMap;</span>
    }

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>