<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>RepeatsParser.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-app</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.app.transform</a> &gt; <span class="el_source">RepeatsParser.java</span></div><h1>RepeatsParser.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.app.transform;

import org.opencb.biodata.models.core.Region;
import org.opencb.biodata.models.variant.avro.Repeat;
import org.opencb.cellbase.app.cli.EtlCommons;
import org.opencb.cellbase.core.serializer.CellBaseFileSerializer;
import org.opencb.commons.ProgressLogger;
import org.opencb.commons.utils.FileUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Created by fjlopez on 05/05/17.
 */
public class RepeatsParser extends CellBaseParser {
    private static final String TRF = &quot;trf&quot;;
    private static final String GSD = &quot;genomicSuperDup&quot;;
    private static final String WM = &quot;windowMasker&quot;;
    private final Path filesDir;

    public RepeatsParser(Path filesDir, CellBaseFileSerializer serializer) {
<span class="fc" id="L25">        super(serializer);</span>
<span class="fc" id="L26">        this.filesDir = filesDir;</span>
<span class="fc" id="L27">    }</span>


    @Override
    public void parse() throws Exception {

<span class="fc" id="L33">        logger.info(&quot;Parsing repeats...&quot;);</span>
<span class="pc bpc" id="L34" title="1 of 2 branches missed.">        if (Files.exists(filesDir.resolve(EtlCommons.TRF_FILE))) {</span>
<span class="fc" id="L35">            parseTrfFile(filesDir.resolve(EtlCommons.TRF_FILE));</span>
        } else {
<span class="nc" id="L37">            logger.warn(&quot;No TRF file found {}&quot;, EtlCommons.TRF_FILE);</span>
<span class="nc" id="L38">            logger.warn(&quot;Skipping TRF file parsing. TRF data models will not be built.&quot;);</span>
        }

<span class="pc bpc" id="L41" title="1 of 2 branches missed.">        if (Files.exists(filesDir.resolve(EtlCommons.GSD_FILE))) {</span>
<span class="fc" id="L42">            parseGsdFile(filesDir.resolve(EtlCommons.GSD_FILE));</span>
        } else {
<span class="nc" id="L44">            logger.warn(&quot;No Genomic Super Duplications file found {}&quot;, EtlCommons.GSD_FILE);</span>
<span class="nc" id="L45">            logger.warn(&quot;Skipping Genomic Super Duplications file parsing. &quot;</span>
                    + &quot;Genomic Super Duplications data models will not be built.&quot;);
        }

<span class="pc bpc" id="L49" title="1 of 2 branches missed.">        if (Files.exists(filesDir.resolve(EtlCommons.WM_FILE))) {</span>
<span class="fc" id="L50">            parseWmFile(filesDir.resolve(EtlCommons.WM_FILE));</span>
        } else {
<span class="nc" id="L52">            logger.warn(&quot;No WindowMasker file found {}&quot;, EtlCommons.WM_FILE);</span>
<span class="nc" id="L53">            logger.warn(&quot;Skipping WindowMasker file parsing. WindowMasker data models will not be built.&quot;);</span>
        }
<span class="fc" id="L55">        logger.info(&quot;Done.&quot;);</span>

<span class="fc" id="L57">    }</span>

    private void parseTrfFile(Path filePath) throws IOException {
<span class="pc" id="L60">        try (BufferedReader bufferedReader = FileUtils.newBufferedReader(filePath)) {</span>
<span class="fc" id="L61">            String line = bufferedReader.readLine();</span>

<span class="fc" id="L63">            ProgressLogger progressLogger = new ProgressLogger(&quot;Parsed TRF lines:&quot;,</span>
<span class="fc" id="L64">                    () -&gt; EtlCommons.countFileLines(filePath), 200).setBatchSize(10000);</span>
<span class="fc bfc" id="L65" title="All 2 branches covered.">            while (line != null) {</span>
<span class="fc" id="L66">                serializer.serialize(parseTrfLine(line));</span>
<span class="fc" id="L67">                line = bufferedReader.readLine();</span>
<span class="fc" id="L68">                progressLogger.increment(1);</span>
            }
<span class="pc bpc" id="L70" title="6 of 8 branches missed.">        }</span>
<span class="fc" id="L71">    }</span>

    private Repeat parseTrfLine(String line) {
<span class="fc" id="L74">        String[] parts = line.split(&quot;\t&quot;);</span>

<span class="fc" id="L76">        return new Repeat(null, Region.normalizeChromosome(parts[1]), Integer.valueOf(parts[2]) + 1,</span>
<span class="fc" id="L77">                Integer.valueOf(parts[3]), Integer.valueOf(parts[5]), Integer.valueOf(parts[7]),</span>
<span class="fc" id="L78">                Float.valueOf(parts[6]), Float.valueOf(parts[8]) / 100, Float.valueOf(parts[10]), parts[16], TRF);</span>
    }

    private void parseGsdFile(Path filePath) throws IOException {
<span class="pc" id="L82">        try (BufferedReader bufferedReader = FileUtils.newBufferedReader(filePath)) {</span>
<span class="fc" id="L83">            String line = bufferedReader.readLine();</span>

<span class="fc" id="L85">            ProgressLogger progressLogger = new ProgressLogger(&quot;Parsed GSD lines:&quot;,</span>
<span class="fc" id="L86">                    () -&gt; EtlCommons.countFileLines(filePath), 200).setBatchSize(10000);</span>
<span class="fc bfc" id="L87" title="All 2 branches covered.">            while (line != null) {</span>
<span class="fc" id="L88">                serializer.serialize(parseGSDLine(line));</span>
<span class="fc" id="L89">                line = bufferedReader.readLine();</span>
<span class="fc" id="L90">                progressLogger.increment(1);</span>
            }
<span class="pc bpc" id="L92" title="6 of 8 branches missed.">        }</span>
<span class="fc" id="L93">    }</span>

    private Repeat parseGSDLine(String line) {
<span class="fc" id="L96">        String[] parts = line.split(&quot;\t&quot;);</span>

<span class="fc" id="L98">        return new Repeat(parts[11], Region.normalizeChromosome(parts[1]), Integer.valueOf(parts[2]) + 1,</span>
<span class="fc" id="L99">                Integer.valueOf(parts[3]), null, null, 2f, Float.valueOf(parts[26]), null,</span>
                null, GSD);

    }

    private void parseWmFile(Path filePath) throws IOException {
<span class="pc" id="L105">        try (BufferedReader bufferedReader = FileUtils.newBufferedReader(filePath)) {</span>
<span class="fc" id="L106">            String line = bufferedReader.readLine();</span>

<span class="fc" id="L108">            ProgressLogger progressLogger = new ProgressLogger(&quot;Parsed WM lines:&quot;,</span>
<span class="fc" id="L109">                    () -&gt; EtlCommons.countFileLines(filePath), 200).setBatchSize(10000);</span>
<span class="fc bfc" id="L110" title="All 2 branches covered.">            while (line != null) {</span>
<span class="fc" id="L111">                serializer.serialize(parseWmLine(line));</span>
<span class="fc" id="L112">                line = bufferedReader.readLine();</span>
<span class="fc" id="L113">                progressLogger.increment(1);</span>
            }
<span class="pc bpc" id="L115" title="6 of 8 branches missed.">        }</span>
<span class="fc" id="L116">    }</span>

    private Repeat parseWmLine(String line) {
<span class="fc" id="L119">        String[] parts = line.split(&quot;\t&quot;);</span>

<span class="fc" id="L121">        return new Repeat(parts[4].replace(&quot;\t&quot;, &quot;&quot;), Region.normalizeChromosome(parts[1]),</span>
<span class="fc" id="L122">                Integer.valueOf(parts[2]) + 1, Integer.valueOf(parts[3]), null, null, null, null, null, null, WM);</span>
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>