<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>GwasParser.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-app</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.app.transform.clinical.variant</a> &gt; <span class="el_source">GwasParser.java</span></div><h1>GwasParser.java</h1><pre class="source lang-java linenums">/*
 * Copyright 2015 OpenCB
 *
 * Licensed under the Apache License, Version 2.0 (the &quot;License&quot;);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an &quot;AS IS&quot; BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.opencb.cellbase.app.transform.clinical.variant;

import htsjdk.tribble.readers.TabixReader;
import org.apache.commons.lang.math.NumberUtils;
import org.opencb.biodata.models.variant.Variant;
import org.opencb.cellbase.app.transform.CellBaseParser;
import org.opencb.cellbase.core.common.clinical.gwas.Gwas;
import org.opencb.cellbase.core.common.clinical.gwas.GwasStudy;
import org.opencb.cellbase.core.common.clinical.gwas.GwasTest;
import org.opencb.cellbase.core.common.clinical.gwas.GwasTrait;
import org.opencb.cellbase.core.serializer.CellBaseSerializer;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.text.NumberFormat;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Luis Miguel Cruz
 * @version 1.2.3
 * @since October 08, 2014
 */
@Deprecated
public class GwasParser extends CellBaseParser {

    private static final int REF = 0;
    private static final int ALT = 1;

    private final Path gwasFile;
    private final Path dbSnpTabixFilePath;

    private int invalidStartRecords;
    private int invalidChromosome;
    private int gwasLinesNotFoundInDbsnp;

    public GwasParser(Path gwasFile, Path dbSnpTabixFilePath, CellBaseSerializer serializer) {
<span class="fc" id="L57">        super(serializer);</span>
<span class="fc" id="L58">        this.gwasFile = gwasFile;</span>
<span class="fc" id="L59">        this.dbSnpTabixFilePath = dbSnpTabixFilePath;</span>
<span class="fc" id="L60">        this.invalidStartRecords = 0;</span>
<span class="fc" id="L61">        this.invalidChromosome = 0;</span>
<span class="fc" id="L62">        this.gwasLinesNotFoundInDbsnp = 0;</span>
<span class="fc" id="L63">    }</span>

    public void parse() {
<span class="pc bpc" id="L66" title="2 of 4 branches missed.">        if (Files.exists(gwasFile) &amp;&amp; Files.exists(dbSnpTabixFilePath)) {</span>
            try {
<span class="fc" id="L68">                logger.info(&quot;Opening gwas file &quot; + gwasFile + &quot; ...&quot;);</span>
<span class="fc" id="L69">                BufferedReader inputReader = new BufferedReader(new FileReader(gwasFile.toFile()));</span>

<span class="fc" id="L71">                logger.info(&quot;Ignoring gwas file header line ...&quot;);</span>
<span class="fc" id="L72">                inputReader.readLine();</span>

<span class="fc" id="L74">                Map&lt;Variant, Gwas&gt; variantMap = new HashMap&lt;&gt;();</span>
<span class="fc" id="L75">                logger.info(&quot;Opening dbSNP tabix file &quot; + dbSnpTabixFilePath + &quot; ...&quot;);</span>
<span class="fc" id="L76">                TabixReader dbsnpTabixReader = new TabixReader(dbSnpTabixFilePath.toString());</span>

<span class="fc" id="L78">                long processedGwasLines = 0;</span>

<span class="fc" id="L80">                logger.info(&quot;Parsing gwas file ...&quot;);</span>
                String line;
<span class="fc bfc" id="L82" title="All 2 branches covered.">                while ((line = inputReader.readLine()) != null) {</span>
<span class="pc bpc" id="L83" title="1 of 2 branches missed.">                    if (!line.isEmpty()) {</span>
<span class="fc" id="L84">                        processedGwasLines++;</span>
<span class="fc" id="L85">                        Gwas gwasRecord = buildGwasObject(line.split(&quot;\t&quot;), dbsnpTabixReader);</span>
<span class="pc bpc" id="L86" title="1 of 2 branches missed.">                        if (gwasRecord != null) {</span>
<span class="fc" id="L87">                            addGwasRecordToVariantMap(variantMap, gwasRecord);</span>
                        }
<span class="fc" id="L89">                    }</span>
                }
<span class="fc" id="L91">                dbsnpTabixReader.close();</span>

<span class="fc" id="L93">                logger.info(&quot;Serializing parsed variants ...&quot;);</span>
<span class="fc bfc" id="L94" title="All 2 branches covered.">                for (Gwas gwasOutputRecord : variantMap.values()) {</span>
<span class="fc" id="L95">                    serializer.serialize(gwasOutputRecord);</span>
<span class="fc" id="L96">                }</span>
<span class="fc" id="L97">                logger.info(&quot;Done&quot;);</span>
<span class="fc" id="L98">                this.disconnect();</span>
<span class="fc" id="L99">                this.printSummary(processedGwasLines, variantMap);</span>


<span class="nc" id="L102">            } catch (IOException e) {</span>
<span class="nc" id="L103">                logger.error(&quot;Unable to parse &quot; + gwasFile + &quot; using dbSNP file &quot; + dbSnpTabixFilePath + &quot;: &quot; + e.getMessage());</span>
<span class="fc" id="L104">            }</span>
        }
<span class="fc" id="L106">    }</span>

    private Gwas buildGwasObject(String[] values, TabixReader dbsnpTabixReader) {
<span class="fc" id="L109">        Gwas gwas = null;</span>
<span class="fc" id="L110">        Integer start = parseStart(values);</span>
<span class="pc bpc" id="L111" title="1 of 2 branches missed.">        if (start != null) {</span>
<span class="fc" id="L112">            Integer end = start;</span>

<span class="fc" id="L114">            String chromosome = parseChromosome(values[11]);</span>
<span class="pc bpc" id="L115" title="1 of 2 branches missed.">            if (chromosome != null) {</span>

<span class="fc" id="L117">                String snpId = values[21].trim();</span>
<span class="fc" id="L118">                String[] refAndAlt = getRefAndAltFromDbsnp(chromosome, start, snpId, dbsnpTabixReader);</span>
<span class="pc bpc" id="L119" title="1 of 2 branches missed.">                if (refAndAlt != null) {</span>

<span class="fc" id="L121">                    gwas = new Gwas(chromosome, start, end, refAndAlt[REF], refAndAlt[ALT], values[10], values[13], values[14],</span>
                            values[15], values[16], values[17], values[18], values[19], values[20], snpId, values[22], values[23],
<span class="fc" id="L123">                            values[24], values[25], parseFloat(values[26]), values[33]);</span>
<span class="fc" id="L124">                    addGwasStudy(values, gwas);</span>

                } else {
<span class="nc" id="L127">                    gwasLinesNotFoundInDbsnp++;</span>
                }
<span class="fc" id="L129">            } else {</span>
<span class="nc" id="L130">                invalidChromosome++;</span>
            }
<span class="fc" id="L132">        } else {</span>
<span class="nc" id="L133">            invalidStartRecords++;</span>
        }

<span class="fc" id="L136">        return gwas;</span>
    }

    private Integer parseStart(String[] values) {
<span class="fc" id="L140">        Integer start = null;</span>
<span class="pc bpc" id="L141" title="1 of 2 branches missed.">        if (NumberUtils.isDigits(values[12])) {</span>
<span class="fc" id="L142">            start = Integer.parseInt(values[12]);</span>
        }
<span class="fc" id="L144">        return start;</span>
    }

    private String parseChromosome(String chromosome) {
<span class="fc" id="L148">        String transformedChromosome = null;</span>
<span class="pc bpc" id="L149" title="1 of 2 branches missed.">        if (!chromosome.isEmpty()) {</span>
<span class="pc bpc" id="L150" title="9 of 14 branches missed.">            switch (chromosome) {</span>
                case &quot;23&quot;:
<span class="fc" id="L152">                    transformedChromosome = &quot;X&quot;;</span>
<span class="fc" id="L153">                    break;</span>
                case &quot;24&quot;:
<span class="nc" id="L155">                    transformedChromosome = &quot;Y&quot;;</span>
<span class="nc" id="L156">                    break;</span>
                case &quot;25&quot;:
<span class="nc" id="L158">                    transformedChromosome = &quot;MT&quot;;</span>
<span class="nc" id="L159">                    break;</span>
                default:
<span class="fc" id="L161">                    transformedChromosome = chromosome;</span>
            }
        }
<span class="fc" id="L164">        return transformedChromosome;</span>
    }

    private Float parseFloat(String value) {
<span class="fc" id="L168">        Float riskAlleleFrequency = null;</span>
<span class="fc bfc" id="L169" title="All 2 branches covered.">        if (NumberUtils.isNumber(value)) {</span>
<span class="fc" id="L170">            riskAlleleFrequency = Float.parseFloat(value);</span>
        }
<span class="fc" id="L172">        return riskAlleleFrequency;</span>
    }

    private String[] getRefAndAltFromDbsnp(String chromosome, Integer start, String snpId, TabixReader dbsnpTabixReader) {
<span class="fc" id="L176">        String[] refAndAlt = null;</span>

<span class="fc" id="L178">        TabixReader.Iterator dbsnpIterator = dbsnpTabixReader.query(chromosome + &quot;:&quot; + start + &quot;-&quot; + start);</span>
        try {
<span class="fc" id="L180">            String dbSnpRecord = dbsnpIterator.next();</span>
<span class="fc" id="L181">            boolean found = false;</span>
<span class="pc bpc" id="L182" title="1 of 4 branches missed.">            while (dbSnpRecord != null &amp;&amp; !found) {</span>
<span class="fc" id="L183">                String[] dbsnpFields = dbSnpRecord.split(&quot;\t&quot;);</span>

<span class="pc bpc" id="L185" title="1 of 2 branches missed.">                if (snpId.equalsIgnoreCase(dbsnpFields[2])) {</span>
<span class="fc" id="L186">                    refAndAlt = new String[2];</span>
<span class="fc" id="L187">                    refAndAlt[REF] = dbsnpFields[3];</span>
<span class="fc" id="L188">                    refAndAlt[ALT] = dbsnpFields[4];</span>
<span class="fc" id="L189">                    found = true;</span>
                }

<span class="fc" id="L192">                dbSnpRecord = dbsnpIterator.next();</span>
<span class="fc" id="L193">            }</span>
<span class="nc" id="L194">        } catch (IOException e) {</span>
<span class="nc" id="L195">            logger.warn(&quot;Error reading position '&quot; + chromosome + &quot;:&quot; + start + &quot;' in dbSNP: &quot; + e.getMessage());</span>
<span class="fc" id="L196">        }</span>

<span class="fc" id="L198">        return refAndAlt;</span>
    }

    private void addGwasStudy(String[] values, Gwas gwas) {
        // Add the study values
<span class="fc" id="L203">        GwasStudy study = new GwasStudy(values[1], values[2], values[3], values[4], values[5], values[6], values[8], values[9], values[32]);</span>
<span class="fc" id="L204">        addGwasTraitToStudy(values, study);</span>
<span class="fc" id="L205">        gwas.addStudy(study);</span>
<span class="fc" id="L206">    }</span>

    private void addGwasTraitToStudy(String[] values, GwasStudy study) {
        // Add the trait values
<span class="fc" id="L210">        GwasTrait trait = new GwasTrait(values[7], values[0]);</span>
<span class="fc" id="L211">        addGwasTestToTrait(values, trait);</span>
<span class="fc" id="L212">        study.addTrait(trait);</span>
<span class="fc" id="L213">    }</span>

    private void addGwasTestToTrait(String[] values, GwasTrait trait) {
        // Add the test values
<span class="fc" id="L217">        Float pValue = parseFloat(values[27]);</span>
<span class="fc" id="L218">        Float pValueMlog = parseFloat(values[28]);</span>
<span class="fc" id="L219">        GwasTest test = new GwasTest(pValue, pValueMlog, values[29], values[30], values[31]);</span>
<span class="fc" id="L220">        trait.addTest(test);</span>
<span class="fc" id="L221">    }</span>

    private void printSummary(long processedGwasLines, Map&lt;Variant, Gwas&gt; variantMap) {
<span class="fc" id="L224">        NumberFormat formatter = NumberFormat.getInstance();</span>
<span class="fc" id="L225">        logger.info(&quot;&quot;);</span>
<span class="fc" id="L226">        logger.info(&quot;Summary&quot;);</span>
<span class="fc" id="L227">        logger.info(&quot;=======&quot;);</span>
<span class="fc" id="L228">        logger.info(&quot;Processed &quot; + formatter.format(processedGwasLines) + &quot; gwas lines&quot;);</span>
<span class="fc" id="L229">        logger.info(&quot;Serialized &quot; + formatter.format(variantMap.size()) + &quot; variants&quot;);</span>
<span class="fc" id="L230">        logger.info(formatter.format(gwasLinesNotFoundInDbsnp) + &quot; gwas lines ignored because variant not found in dbsnp&quot;);</span>
<span class="pc bpc" id="L231" title="1 of 2 branches missed.">        if (invalidStartRecords != 0) {</span>
<span class="nc" id="L232">            logger.info(formatter.format(invalidStartRecords) + &quot; gwas lines ignored because have no valid 'start' value&quot;);</span>
        }
<span class="pc bpc" id="L234" title="1 of 2 branches missed.">        if (invalidChromosome != 0) {</span>
<span class="nc" id="L235">            logger.info(formatter.format(invalidChromosome) + &quot; gwas lines ignored because have no valid chromosome&quot;);</span>
        }
<span class="fc" id="L237">    }</span>

    private void addGwasRecordToVariantMap(Map&lt;Variant, Gwas&gt; variantMap, Gwas gwasRecord) {
<span class="fc" id="L240">        String[] alternates = gwasRecord.getAlternate().split(&quot;,&quot;);</span>
<span class="fc bfc" id="L241" title="All 2 branches covered.">        for (int i = 0; i &lt; alternates.length; i++) {</span>
<span class="fc" id="L242">            String alternate = alternates[i];</span>
<span class="fc" id="L243">            Variant variantKey = new Variant(</span>
<span class="fc" id="L244">                    gwasRecord.getChromosome(), gwasRecord.getStart(), gwasRecord.getEnd(), gwasRecord.getReference(), alternate);</span>
<span class="fc bfc" id="L245" title="All 2 branches covered.">            if (variantMap.containsKey(variantKey)) {</span>
<span class="fc" id="L246">                updateGwasEntry(variantMap, gwasRecord, variantKey);</span>
            } else {
                // if a gwas record has several alternatives, it has to be cloned to avoid side effects (set gwasRecord
                // alternative would update the previous instance of gwas record saved in the 'variantMap')
<span class="fc" id="L250">                gwasRecord = cloneGwasRecordIfNecessary(gwasRecord, i);</span>
<span class="fc" id="L251">                gwasRecord.setAlternate(alternate);</span>
<span class="fc" id="L252">                variantMap.put(variantKey, gwasRecord);</span>
            }
        }
<span class="fc" id="L255">    }</span>

    private Gwas cloneGwasRecordIfNecessary(Gwas gwasRecord, int i) {
<span class="fc bfc" id="L258" title="All 2 branches covered.">        if (i &gt; 0) {</span>
<span class="fc" id="L259">            gwasRecord = new Gwas(gwasRecord);</span>
        }
<span class="fc" id="L261">        return gwasRecord;</span>
    }

    private void updateGwasEntry(Map&lt;Variant, Gwas&gt; variantMap, Gwas gwasVO, Variant gwasKey) {
<span class="fc" id="L265">        Gwas gwas = variantMap.get(gwasKey);</span>
<span class="fc" id="L266">        gwas.addStudies(gwasVO.getStudies());</span>
<span class="fc" id="L267">        variantMap.put(gwasKey, gwas);</span>
<span class="fc" id="L268">    }</span>
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>