<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>HgvsSNVCalculator.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">cellbase-core</a> &gt; <a href="index.source.html" class="el_package">org.opencb.cellbase.core.variant.annotation.hgvs</a> &gt; <span class="el_source">HgvsSNVCalculator.java</span></div><h1>HgvsSNVCalculator.java</h1><pre class="source lang-java linenums">package org.opencb.cellbase.core.variant.annotation.hgvs;

import org.opencb.biodata.models.core.Transcript;
import org.opencb.biodata.models.variant.Variant;
import org.opencb.cellbase.core.api.GenomeDBAdaptor;
import org.opencb.cellbase.core.variant.annotation.VariantAnnotationUtils;

import java.util.Collections;
import java.util.List;

/**
 * Created by fjlopez on 15/06/17.
 */
public class HgvsSNVCalculator extends HgvsCalculator {
    public HgvsSNVCalculator(GenomeDBAdaptor genomeDBAdaptor) {
<span class="nc" id="L16">        super(genomeDBAdaptor);</span>
<span class="nc" id="L17">    }</span>

    @Override
    protected List&lt;String&gt; run(Variant variant, Transcript transcript, String geneId, boolean normalize) {
<span class="nc" id="L21">        Variant normalizedVariant = normalize(variant, normalize);</span>
<span class="nc" id="L22">        return calculateTranscriptHgvs(normalizedVariant, transcript, geneId);</span>
    }

    /**
     * Generates cdna HGVS names from an SNV.
     * @param transcript Transcript object that will be used as a reference
     */
    private List&lt;String&gt; calculateTranscriptHgvs(Variant variant, Transcript transcript, String geneId) {

<span class="nc" id="L31">        String mutationType = &quot;&gt;&quot;;</span>

        // Populate HGVSName parse tree.
<span class="nc" id="L34">        HgvsStringBuilder hgvsStringBuilder = new HgvsStringBuilder();</span>

        // Populate coordinates.
        // Use cDNA coordinates.
<span class="nc bnc" id="L38" title="All 2 branches missed.">        hgvsStringBuilder.setKind(isCoding(transcript) ? HgvsStringBuilder.Kind.CODING : HgvsStringBuilder.Kind.NON_CODING);</span>

<span class="nc" id="L40">        hgvsStringBuilder.setCdnaStart(genomicToCdnaCoord(transcript, variant.getStart()));</span>
<span class="nc" id="L41">        hgvsStringBuilder.setCdnaEnd(hgvsStringBuilder.getCdnaStart());</span>

        // Populate prefix.
<span class="nc" id="L44">        hgvsStringBuilder.setTranscriptId(transcript.getId());</span>
<span class="nc" id="L45">        hgvsStringBuilder.setGeneId(geneId);</span>

        String reference;
        String alternate;
        // Convert alleles to transcript strand.
<span class="nc bnc" id="L50" title="All 2 branches missed.">        if (transcript.getStrand().equals(&quot;-&quot;)) {</span>
<span class="nc" id="L51">            reference = String.valueOf(VariantAnnotationUtils.COMPLEMENTARY_NT.get(variant.getReference().charAt(0)));</span>
<span class="nc" id="L52">            alternate = String.valueOf(VariantAnnotationUtils.COMPLEMENTARY_NT.get(variant.getAlternate().charAt(0)));</span>
        } else {
<span class="nc" id="L54">            reference = variant.getReference();</span>
<span class="nc" id="L55">            alternate = variant.getAlternate();</span>
        }

        // Populate alleles.
<span class="nc" id="L59">        hgvsStringBuilder.setMutationType(mutationType);</span>
<span class="nc" id="L60">        hgvsStringBuilder.setReference(reference);</span>
<span class="nc" id="L61">        hgvsStringBuilder.setAlternate(alternate);</span>

<span class="nc" id="L63">        return Collections.singletonList(hgvsStringBuilder.format());</span>
    }

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.7.201606060606</span></div></body></html>